File: typescript/packages/mechanisms/evm/test/unit/assetCache.test.ts
import { describe, it, expect, beforeEach } from "vitest";
import type { FacilitatorEvmSigner } from "../../src/signer";
import { ErrAssetNotDeployedContract } from "../../src/exact/facilitator/errors";
import {
  DEFAULT_ASSET_CONTRACT_CACHE_TTL_MS,
  MAX_ASSET_CONTRACT_CACHE_ENTRIES,
  globalAssetContractCache,
  resetAssetContractCache,
  startAssetContractCheck,
  type AssetContractCacheKey,
} from "../../src/assetCache";

const cacheTestAsset = "0x00000000000000000000000000000000000000bb";

function countingCodeSigner(): { signer: FacilitatorEvmSigner; calls: { count: number } } {
  const calls = { count: 0 };
  const signer: FacilitatorEvmSigner = {
    getAddresses: () => [],
    readContract: async () => 0n,
    verifyTypedData: async () => false,
    writeContract: async () => "0x",
    sendTransaction: async () => "0x",
    waitForTransactionReceipt: async () => ({ status: "success" }),
    getCode: async () => {
      calls.count += 1;
      return "0x6060";
    },
  };
  return { signer, calls };
}

describe("asset contract cache", () => {
  beforeEach(() => {
    resetAssetContractCache();
  });

  it("caches a positive result so only the first check reaches the RPC", async () => {
    const { signer, calls } = countingCodeSigner();
    for (let i = 0; i < 3; i++) {
      const reason = await startAssetContractCheck(signer, "eip155:84532", cacheTestAsset).await();
      expect(reason).toBe("");
    }

    expect(calls.count).toBe(1);
  });

  it("does not cache when the caller never awaits", async () => {
    const { signer, calls } = countingCodeSigner();
    const abandoned = startAssetContractCheck(signer, "eip155:84532", cacheTestAsset);
    await abandoned.results;

    const reason = await startAssetContractCheck(signer, "eip155:84532", cacheTestAsset).await();
    expect(reason).toBe("");
    expect(calls.count).toBe(2);
  });

  it("does not cache a negative result so a later deploy is visible", async () => {
    let code: `0x${string}` = "0x";
    const calls = { count: 0 };
    const signer: FacilitatorEvmSigner = {
      getAddresses: () => [],
      readContract: async () => 0n,
      verifyTypedData: async () => false,
      writeContract: async () => "0x",
      sendTransaction: async () => "0x",
      waitForTransactionReceipt: async () => ({ status: "success" }),
      getCode: async () => {
        calls.count += 1;
        return code;
      },
    };

    const first = await startAssetContractCheck(signer, "eip155:84532", cacheTestAsset).await();
    expect(first).toBe(ErrAssetNotDeployedContract);
    expect(calls.count).toBe(1);

    code = "0x6060";
    const second = await startAssetContractCheck(signer, "eip155:84532", cacheTestAsset).await();
    expect(second).toBe("");
    expect(calls.count).toBe(2);

    const third = await startAssetContractCheck(signer, "eip155:84532", cacheTestAsset).await();
    expect(third).toBe("");
    expect(calls.count).toBe(2);
  });

  it("skips the cache when the network is empty", async () => {
    const { signer, calls } = countingCodeSigner();
    for (let i = 0; i < 2; i++) {
      await startAssetContractCheck(signer, "", cacheTestAsset).await();
    }

    expect(calls.count).toBe(2);
    expect(
      globalAssetContractCache.isFresh({ network: "", asset: cacheTestAsset }, Date.now()),
    ).toBe(false);
  });

  it("expires entries after the TTL", () => {
    const key: AssetContractCacheKey = { network: "eip155:84532", asset: cacheTestAsset };
    const start = Date.now();
    globalAssetContractCache.record(key, start);

    expect(
      globalAssetContractCache.isFresh(key, start + DEFAULT_ASSET_CONTRACT_CACHE_TTL_MS - 1000),
    ).toBe(true);
    expect(
      globalAssetContractCache.isFresh(key, start + DEFAULT_ASSET_CONTRACT_CACHE_TTL_MS + 1000),
    ).toBe(false);
  });

  it("is bounded so many distinct deployed contracts cannot grow it without limit", () => {
    const now = Date.now();
    for (let i = 0; i < MAX_ASSET_CONTRACT_CACHE_ENTRIES + 500; i++) {
      globalAssetContractCache.record(
        { network: "eip155:84532", asset: `0x${i.toString(16).padStart(40, "0")}` },
        now,
      );
    }

    expect(globalAssetContractCache.expiries.size).toBeLessThanOrEqual(
      MAX_ASSET_CONTRACT_CACHE_ENTRIES,
    );
  });
});

File: go/mechanisms/evm/asset_cache_test.go
package evm

import (
	"context"
	"fmt"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

const cacheTestAsset = "0x00000000000000000000000000000000000000bb"

// countingCodeSigner reports the asset as deployed and counts eth_getCode calls.
type countingCodeSigner struct {
	mockStrictSigner
	calls int
}

func (s *countingCodeSigner) GetCode(_ context.Context, _ string) ([]byte, error) {
	s.calls++
	return []byte{0x60, 0x60}, nil
}

func TestAssetContractCheckCachesPositiveResult(t *testing.T) {
	ResetAssetContractCache()

	signer := &countingCodeSigner{}
	for range 3 {
		reason, err := StartAssetContractCheck(context.Background(), signer, "eip155:84532", cacheTestAsset).Await()
		require.NoError(t, err)
		assert.Empty(t, reason)
	}

	assert.Equal(t, 1, signer.calls, "only the first check should reach the RPC")
}

// A caller that never awaits must leave the cache untouched, so cache contents do not depend on
// goroutine scheduling.
func TestAssetContractCheckWithoutAwaitDoesNotCache(t *testing.T) {
	ResetAssetContractCache()

	signer := &countingCodeSigner{}
	abandoned := StartAssetContractCheck(context.Background(), signer, "eip155:84532", cacheTestAsset)
	<-abandoned.results

	reason, err := StartAssetContractCheck(context.Background(), signer, "eip155:84532", cacheTestAsset).Await()
	require.NoError(t, err)
	assert.Empty(t, reason)
	assert.Equal(t, 2, signer.calls, "the abandoned check must not have populated the cache")
}

// An empty network cannot be cached: entries would otherwise collide across chains, where one
// address can hold bytecode on one chain and nothing on another.
func TestAssetContractCacheSkipsEmptyNetwork(t *testing.T) {
	ResetAssetContractCache()

	signer := &countingCodeSigner{}
	for range 2 {
		_, err := StartAssetContractCheck(context.Background(), signer, "", cacheTestAsset).Await()
		require.NoError(t, err)
	}

	assert.Equal(t, 2, signer.calls, "an empty network must bypass the cache")
	assert.False(t, globalAssetContractCache.isFresh(assetContractCacheKey{asset: cacheTestAsset}, time.Now()))
}

func TestAssetContractCacheEntriesExpire(t *testing.T) {
	ResetAssetContractCache()

	key := assetContractCacheKey{network: "eip155:84532", asset: cacheTestAsset}
	start := time.Now()
	globalAssetContractCache.record(key, start)

	assert.True(t, globalAssetContractCache.isFresh(key, start.Add(DefaultAssetContractCacheTTL-time.Second)))
	assert.False(t, globalAssetContractCache.isFresh(key, start.Add(DefaultAssetContractCacheTTL+time.Second)))
}

// The cache is keyed partly by caller-supplied asset addresses, so it must not grow without
// bound when many distinct deployed contracts are named within one TTL window.
func TestAssetContractCacheIsBounded(t *testing.T) {
	ResetAssetContractCache()

	now := time.Now()
	for i := range maxAssetContractCacheEntries + 500 {
		globalAssetContractCache.record(
			assetContractCacheKey{network: "eip155:84532", asset: fmt.Sprintf("0x%040x", i)},
			now,
		)
	}

	globalAssetContractCache.mu.RLock()
	size := len(globalAssetContractCache.expiries)
	globalAssetContractCache.mu.RUnlock()

	assert.LessOrEqual(t, size, maxAssetContractCacheEntries)
}
